esp_commands/src/esp_commands.c

```diff
@@ -0,0 +1,761 @@
+     * FOR_EACH_XX_COMMAND will not go through the whole list of static (resp. dynamic) commands but
+     * through the empty set, so no command will be walked.
+     */
+
+    esp_command_set_t *static_set = cmd_sets ? &cmd_sets->static_set : NULL;
+    /* it is possible that the set is empty, in which case set static_set to NULL
+     * to prevent the for loop to try to access a list of commands pointer set to NULL */
+    if (static_set && !static_set->cmd_ptr_set) {
+        static_set = NULL;
+    }
+    FOR_EACH_STATIC_COMMAND(static_set, cmd) {
+        continue_walk = cmd_walker(cmd_walker_ctx, cmd);
+        if (!continue_walk) {
+            return;
+        }
+    }
+
+    esp_command_set_t *dynamic_set = cmd_sets ? &cmd_sets->dynamic_set : NULL;
+    /* it is possible that the set is empty, in which case set dynamic_set to NULL
+     * to prevent the for loop to try to access a list of commands pointer set to NULL */
+    if (dynamic_set && !dynamic_set->cmd_ptr_set) {
+        dynamic_set = NULL;
+    }
+    esp_dynamic_commands_lock();
+    FOR_EACH_DYNAMIC_COMMAND(dynamic_set, cmd) {
+        continue_walk = cmd_walker(cmd_walker_ctx, cmd);
+        if (!continue_walk) {
+            esp_dynamic_commands_unlock();
+            return;
+        }
+    }
+    esp_dynamic_commands_unlock();
+}
+
+typedef struct find_cmd_ctx {
+    const char *name; /*!< the name to check commands against */
+    esp_command_t *cmd; /*!< the command matching the name */
+} find_cmd_ctx_t;
+
+static inline __attribute__((always_inline))
+bool compare_command_name(void *ctx, esp_command_t *cmd)
+{
+    /* called by esp_commands_find_command through go_through_commands,
+     * ctx cannot be NULL */
+    find_cmd_ctx_t *cmd_ctx = (find_cmd_ctx_t *)ctx;
+
+    /* called by go_through_commands, thus cmd cannot be NULL */
+    if (strcmp(cmd->name, cmd_ctx->name) == 0) {
+        /* command found, store it in the ctx so esp_commands_find_command
+         * can process it. Notify go_through_commands to stop the walk by
+         * returning false */
+        cmd_ctx->cmd = cmd;
+        return false;
+    }
+
+    /* command not matching with the name from the ctx, continue the walk */
+    return true;
+}
+
+void *esp_commands_malloc(const size_t malloc_size)
+{
+    xSemaphoreTake(s_esp_commands_mutex, portMAX_DELAY);
+    const uint32_t caps = s_config.heap_caps_used;
+    xSemaphoreGive(s_esp_commands_mutex);
+
+    return heap_caps_malloc(malloc_size, caps);
+}
+
+esp_err_t esp_commands_update_config(const esp_commands_config_t *config)
+{
+    if (!config ||
+            (config->max_cmdline_args == 0) ||
+            (config->max_cmdline_length == 0)) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    /* Initialize the mutex if not initialized. Since this function is not mandatory
+     * to be called for the component to work properly, it is possible that the mutex
+     * was never initialized before. */
```

Comment:
Will the mutex behave as expected if the user doesnt call `esp_commands_update_config` but instead directly call execute (or other functions that can take the mutex)?